File: mssp/examples/019-applicability-is-part-of-the-answer/src/island_test.py
"""The island test.

    python src/island_test.py

Section 3 is the control: a MEASURED zero, which is what stops zero from being
a synonym for "nothing to measure". Section 4 proves the unmeasurable unit's
stated reason by running it, rather than believing the sentence.
"""
import json
import pathlib
import re
import sys

from SCL import policy
from SMS import measure
from TMS.sources import declaration_is_baked_in, declares_openly, never_declares
from TMS.policies import ignore_declared, retry_declared

HERE = pathlib.Path(__file__).parent
CONTRACT = json.loads((HERE / "FMS" / "contract.json").read_text(encoding="utf-8"))
FAILURES = []
RAN = [0]


def check(label, ok, detail=""):
    RAN[0] += 1
    print(f'  {"PASS" if ok else "FAIL"}  {label}{f" - {detail}" if detail else ""}')
    if not ok:
        FAILURES.append(label)


sources, policies, problems = measure.load()
under_retry = {m["source"]: m for m in measure.measure_all(retry_declared, sources)}
under_ignore = {m["source"]: m for m in measure.measure_all(ignore_declared, sources)}

print("\n== 1. every unit is an island, and FMS matches the tree")
check("loading raised no problems", not problems, "; ".join(problems))
for unit, declared in CONTRACT["units"].items():
    directory = HERE.joinpath(*unit.split("/"))
    on_disk = sorted(p.name for p in directory.glob("*.py") if p.name != "__init__.py")
    check(f"{unit}: FMS declares what is on disk", on_disk == sorted(declared),
          f'disk {", ".join(on_disk)} | FMS {", ".join(sorted(declared))}')
    for name in on_disk:
        body = directory.joinpath(name).read_text(encoding="utf-8")
        siblings = [n[:-3] for n in on_disk if n != name]
        check(f"{unit}/{name} reaches no sibling",
              not [s for s in siblings if re.search(rf"\bimport\b.*\b{s}\b", body)])
        check(f"{unit}/{name} reaches no other set",
              not re.search(r"\b(from|import)\s+(SMS|DMS|SCL|FMS)\b", body))

print("\n== 2. a measurement carries its applicability")
for name, m in sorted(under_retry.items()):
    check(f"{name} says whether it was measured", isinstance(m["applicable"], bool))
check("a non-applicable reading carries no value", under_retry["baked-in"]["value"] is None)
check("and carries a reason instead", bool(under_retry["baked-in"]["reason"]))
check("an applicable reading carries a value and no reason",
      under_retry["declares-openly"]["value"] == 3 and under_retry["declares-openly"]["reason"] is None)
check("NOT_APPLICABLE is not a number", not isinstance(measure.NOT_APPLICABLE, (int, float)))

print("\n== 3. the control - a MEASURED zero")
control = under_ignore["declares-openly"]
absent = under_retry["baked-in"]
check("under ignore-declared, declares-openly measures zero", control["value"] == 0)
check("and it WAS measured - both arms ran", control["applicable"] is True)
check("and they agree", control["declared"] == control["suppressed"], f'{control["declared"]}')
check("baked-in also reads as nothing", absent["value"] is None)
check("but it was not measured", absent["applicable"] is False)
check("so a bare 0 cannot separate them, and applicability can",
      control["applicable"] != absent["applicable"] and (control["value"] or 0) == 0)
check("never-declares gives a measured zero too, under both policies",
      under_retry["never-declares"]["value"] == 0 and under_ignore["never-declares"]["value"] == 0)

print("\n== 4. the unmeasurable unit's reason, proved by running it")
with_declaration = measure.run_one(declaration_is_baked_in, 1)
without_declaration = measure.run_one(declaration_is_baked_in, 1, suppress_declaration=True)
check("suppressing the field does not remove the marker record",
      any("marker" in r for r in without_declaration["records"]))
check("so the two arms hold a different number of records than a clean unit's arms do",
      len(with_declaration["records"]) == len(without_declaration["records"])
      and any("marker" in r for r in with_declaration["records"]))
clean_with = measure.run_one(declares_openly, 1)
clean_without = measure.run_one(declares_openly, 1, suppress_declaration=True)
check("a suppressible unit's two arms are record-identical",
      clean_with["records"] == clean_without["records"])
check("and differ only in the declaration field",
      clean_with["incomplete_because"] is not None and clean_without["incomplete_because"] is None)
check("baked-in's declaration survives suppression, which is why there is no second arm",
      any(r.get("marker") for r in without_declaration["records"]))
check("and the unit says so itself, up front",
      declaration_is_baked_in.DECLARATION_IS_SUPPRESSIBLE is False)

print("\n== 5. the total refuses instead of defaulting to zero")
try:
    measure.total_incentive(list(under_retry.values()))
    check("a total across an unmeasured unit is refused", False, "it returned a number")
except ValueError as refused:
    check("a total across an unmeasured unit is refused", True, str(refused))
    check("and the refusal names the unit", "baked-in" in str(refused))
measurable = [m for m in under_retry.values() if m["applicable"]]
check("a total across measured units alone is allowed",
      measure.total_incentive(measurable) == 3, f"{measure.total_incentive(measurable)}")
check("which is a different number from the one a silent skip would print",
      measure.total_incentive(measurable) == 3 and len(measurable) < len(under_retry))

print("\n== 6. SCL can only be contradicted by something that was measured")
check("a measured positive contradicts", policy.contradicts(under_retry["declares-openly"]) is not None)
check("a measured zero does not", policy.contradicts(under_retry["never-declares"]) is None)
check("and a non-applicable reading does not either",
      policy.contradicts(under_retry["baked-in"]) is None)
check("but it must not read as agreement - main names it separately",
      "NOT MEASURED" in (HERE / "main.py").read_text(encoding="utf-8"))
check("the deployment calls an unmeasurable unit fatal", policy.unmeasurable_is_fatal())

print("\n== 7. the guards from 013, 015 and 018 still hold")


def drill_source(name, **attributes):
    module = type(sys)(name)
    module.NAME = name
    module.CAN_FAIL_WITH = ["x"]
    module.HELD = 1
    module.DECLARATION_IS_SUPPRESSIBLE = True
    module.collect = lambda budget=1: {"records": [{"from": name, "id": "d-1"}],
                                       "incomplete_because": None}
    for key, value in attributes.items():
        setattr(module, key, value)
    return module


_, _, mute = measure.load(extra_sources=[drill_source("drill-mute", CAN_FAIL_WITH=[])])
check("DRILL: an empty CAN_FAIL_WITH is refused", any("CAN_FAIL_WITH is empty" in p for p in mute))
_, _, vouch = measure.load(extra_sources=[drill_source("drill-vouch", COMPLETE=True)])
check("DRILL: a source declaring COMPLETE is refused", any("declares COMPLETE" in p for p in vouch))
silent = drill_source("drill-unsaid")
del silent.DECLARATION_IS_SUPPRESSIBLE
_, _, unsaid = measure.load(extra_sources=[silent])
check("DRILL: a source that will not say whether it is suppressible is refused",
      any("does not say whether its declaration can be suppressed" in p for p in unsaid))
secretive = type(sys)("drill_secretive")
secretive.POLICY = "drill-secretive"
secretive.apply = lambda runs, rerun: [{**r, "kept": r["records"]} for r in runs]
_, _, told = measure.load(extra_policies=[secretive])
check("DRILL: a policy that will not say what it does with a declaration is refused",
      any("does not say what it does" in p for p in told))
_, _, nameless = measure.load(extra_policies=[type(sys)("drill_nameless")])
check("DRILL: a policy module with no POLICY is refused by name, not by crashing",
      any("does not declare POLICY" in p for p in nameless))
_, _, honest = measure.load(extra_sources=[drill_source("drill-honest")])
check("and an honest unit raises nothing", not honest)

print("\n== 8. what this still cannot do, asserted so it stays measured")
check("nothing here reads intent",
      not re.search(r"\bintent\b|\bstrategic\b", (HERE / "SMS" / "measure.py").read_text(encoding="utf-8")))
check("the reason for non-applicability is the unit's own word, checked only by the record shape",
      declaration_is_baked_in.DECLARATION_IS_SUPPRESSIBLE is False
      and any(r.get("marker") for r in without_declaration["records"]),
      "a unit could declare itself unsuppressible and be lying; that reads as n/a, not as zero")
check("the counterfactual is still one unit, one policy, one run's data",
      under_retry["declares-openly"]["value"] != under_ignore["declares-openly"]["value"],
      f'{under_retry["declares-openly"]["value"]} under retry, '
      f'{under_ignore["declares-openly"]["value"]} under ignore - same unit, same data')

print()
if FAILURES:
    print(f'  {len(FAILURES)} FAILED: {" | ".join(FAILURES)}')
    sys.exit(1)
print(f"  {RAN[0]} checks passed - {len(sources)} sources, {len(policies)} policies, "
      f"3 kinds of answer")

File: mssp/examples/018-who-the-declaration-serves/src/island_test.mjs
// The island test.
//
//   node src/island_test.mjs
//
// Section 2 is the control that makes everything after it mean something.
// Section 4 is the finding. Section 6 proves the contradiction check reads SCL
// rather than a constant.
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import * as pipeline from "./SMS/pipeline.mjs";
import * as policy from "./SCL/policy.mjs";
import * as honestPage from "./TMS/sources/honest_page.mjs";
import * as silentPage from "./TMS/sources/silent_page.mjs";
import * as refuseDeclared from "./TMS/policies/refuse_declared.mjs";
import * as retryDeclared from "./TMS/policies/retry_declared.mjs";

const here = path.dirname(fileURLToPath(import.meta.url));
const CONTRACT = JSON.parse(fs.readFileSync(path.join(here, "FMS", "contract.json"), "utf8"));
const failures = [];
let ran = 0;
const check = (label, ok, detail = "") => {
  ran += 1;
  process.stdout.write(`  ${ok ? "PASS" : "FAIL"}  ${label}${detail ? ` - ${detail}` : ""}\n`);
  if (!ok) failures.push(label);
};
const say = (line = "") => process.stdout.write(`${line}\n`);

const { sources, policies, problems } = pipeline.load();
const incentiveUnder = (module, name) => pipeline.incentive(module, sources, name);

say("\n== 1. every unit is an island, and FMS matches the tree");
check("loading raised no problems", problems.length === 0, problems.join("; "));
for (const [unit, declared] of Object.entries(CONTRACT.units)) {
  const dir = path.join(here, ...unit.split("/"));
  const onDisk = fs.readdirSync(dir).filter((n) => n.endsWith(".mjs")).sort();
  check(`${unit}: FMS declares what is on disk`,
    JSON.stringify(onDisk) === JSON.stringify([...declared].sort()),
    `disk ${onDisk.join(", ")} | FMS ${[...declared].sort().join(", ")}`);
  for (const file of onDisk) {
    const body = fs.readFileSync(path.join(dir, file), "utf8");
    const siblings = onDisk.filter((n) => n !== file).map((n) => n.replace(/\.mjs$/, ""));
    check(`${unit}/${file} reaches no sibling`,
      !siblings.some((s) => new RegExp(`from\\s+["'][^"']*${s}`).test(body)));
    check(`${unit}/${file} reaches no other set`,
      !/from\s+["'][^"']*\/(SMS|DMS|SCL|FMS)\//.test(body));
  }
}

say("\n== 2. the control - two units that differ only in the declaration");
check("honest-page and silent-page hold the same number of records",
  honestPage.HELD === silentPage.HELD, `${honestPage.HELD} and ${silentPage.HELD}`);
for (const budget of [1, 2, 3]) {
  check(`at budget ${budget} they hand over the same number`,
    honestPage.collect({ budget }).records.length === silentPage.collect({ budget }).records.length,
    `${honestPage.collect({ budget }).records.length}`);
}
check("and at budget 1 exactly one of them declares",
  Boolean(honestPage.collect({ budget: 1 }).incomplete_because)
  !== Boolean(silentPage.collect({ budget: 1 }).incomplete_because));
check("so any difference between them is attributable to the declaration alone",
  honestPage.HELD === silentPage.HELD
  && honestPage.collect({ budget: 1 }).records.length === silentPage.collect({ budget: 1 }).records.length);

say("\n== 3. under refuse-declared, declaring costs the unit");
const refusedHonest = incentiveUnder(refuseDeclared, "honest-page");
const refusedSilent = incentiveUnder(refuseDeclared, "silent-page");
check("honest-page contributes nothing once it declares", refusedHonest.declared === 0);
check("and would have contributed 3 with the declaration suppressed", refusedHonest.suppressed === 3);
check("so the incentive is negative", refusedHonest.delta < 0, `${refusedHonest.delta}`);
check("the control moves by nothing", refusedSilent.delta === 0,
  `${refusedSilent.suppressed} -> ${refusedSilent.declared}`);
check("which is what says the machinery is not just moving numbers",
  refusedSilent.delta === 0 && refusedHonest.delta !== 0);

say("\n== 4. under retry-declared, the same declaration pays");
const retriedHonest = incentiveUnder(retryDeclared, "honest-page");
const retriedSilent = incentiveUnder(retryDeclared, "silent-page");
check("honest-page contributes 6 once it declares", retriedHonest.declared === 6);
check("and would have contributed 3 with the declaration suppressed", retriedHonest.suppressed === 3);
check("so the incentive is POSITIVE", retriedHonest.delta > 0, `+${retriedHonest.delta}`);
check("the control still moves by nothing", retriedSilent.delta === 0);
const applied = pipeline.through(retryDeclared, sources);
check("and the honest unit ends up AHEAD of the silent one holding identical data",
  pipeline.contribution(applied, "honest-page") > pipeline.contribution(applied, "silent-page"),
  `${pipeline.contribution(applied, "honest-page")} vs ${pipeline.contribution(applied, "silent-page")}`);
check("the same declaration points opposite ways under the two policies",
  Math.sign(refusedHonest.delta) === -Math.sign(retriedHonest.delta));
check("and neither policy withheld what it does with a declaration",
  Boolean(refuseDeclared.WHAT_IT_DOES_WITH_A_DECLARATION)
  && Boolean(retryDeclared.WHAT_IT_DOES_WITH_A_DECLARATION));

say("\n== 5. FMS's numbers are the run's numbers");
const fmsRefuse = CONTRACT.the_measurement.under_refuse_declared;
const fmsRetry = CONTRACT.the_measurement.under_retry_declared;
check("FMS's refuse-declared figures match the pipeline",
  fmsRefuse["honest-page"] === refusedHonest.delta && fmsRefuse["silent-page"] === refusedSilent.delta,
  `${JSON.stringify(fmsRefuse)} vs ${refusedHonest.delta}/${refusedSilent.delta}`);
check("FMS's retry-declared figures match the pipeline",
  fmsRetry["honest-page"] === retriedHonest.delta && fmsRetry["silent-page"] === retriedSilent.delta,
  `${JSON.stringify(fmsRetry)} vs ${retriedHonest.delta}/${retriedSilent.delta}`);

say("\n== 6. SCL states the assumption, and the check reads SCL");
const inForce = policies[policy.policyName()];
const measuredInForce = { source: "honest-page", ...incentiveUnder(inForce, "honest-page") };
check("the deployment in force is retry-declared", policy.policyName() === "retry-declared");
check("SCL writes down what it assumes", policy.assumption() === "self-penalising");
check("and the measurement contradicts it", policy.contradicts(measuredInForce) !== null,
  policy.contradicts(measuredInForce) ?? "");
check("a negative incentive does NOT contradict it",
  policy.contradicts({ source: "x", declared: 0, suppressed: 3, delta: -3 }) === null);
check("and neither does zero",
  policy.contradicts({ source: "x", declared: 3, suppressed: 3, delta: 0 }) === null);
check("the deployment calls a contradiction fatal", policy.isFatal());

say("\n== 7. the guards from 013 and 015 still hold");
const vouching = {
  NAME: "drill-vouch", CAN_FAIL_WITH: ["x"], HELD: 1, COMPLETE: true,
  collect: () => ({ records: [], incomplete_because: null }),
};
check("DRILL: a source declaring COMPLETE is still refused",
  pipeline.load([vouching]).problems.some((p) => /declares COMPLETE/.test(p)));
const mute = {
  NAME: "drill-mute", CAN_FAIL_WITH: [], HELD: 1,
  collect: () => ({ records: [], incomplete_because: null }),
};
check("DRILL: an empty CAN_FAIL_WITH is still refused",
  pipeline.load([mute]).problems.some((p) => /CAN_FAIL_WITH is empty/.test(p)));
const secretive = {
  POLICY: "drill-secretive",
  apply: (runs) => runs.map((r) => ({ ...r, kept: r.records, note: null })),
};
check("DRILL: a policy that will not say what it does with a declaration is refused",
  pipeline.load([], [secretive]).problems.some((p) => /does not say what it does/.test(p)));

say("\n== 8. what the measurement cannot do, asserted so it stays measured");
const neverDeclares = {
  NAME: "drill-never", CAN_FAIL_WITH: ["x"], HELD: 2,
  collect: () => ({ records: [{ from: "drill-never", id: "n-1" }], incomplete_because: null }),
};
const withExtra = pipeline.load([neverDeclares]);
const flat = pipeline.incentive(retryDeclared, withExtra.sources, "drill-never");
check("a unit that never declares measures as zero", flat.delta === 0,
  `${flat.suppressed} -> ${flat.declared}`);
check("and so does the control, so a zero reading does not say WHICH it is",
  flat.delta === refusedSilent.delta,
  "the limit named in FMS non_goals - if this goes red, the limit changed and the text must too");
check("nothing here reads intent - the reading is outcome only",
  !/\bintent\b|\bdeliberate\b|\bstrategic\b/
    .test(fs.readFileSync(path.join(here, "SMS", "pipeline.mjs"), "utf8")));
check("and the report never prints a delta without the pair it came from",
  /suppressed/.test(fs.readFileSync(path.join(here, "DMS", "report.mjs"), "utf8"))
  && /declared/.test(fs.readFileSync(path.join(here, "DMS", "report.mjs"), "utf8")));

say("");
if (failures.length > 0) {
  say(`  ${failures.length} FAILED: ${failures.join(" | ")}`);
  process.exitCode = 1;
} else {
  say(`  ${ran} checks passed - ${Object.keys(sources).length} sources, ${Object.keys(policies).length} policies`);
}
